### ISSUE TITLE
New buzzwords Nuclio and the Future of Serverless Computing

### ISSUE THREAD

**Issue by postmillenial** [postmillenial/reading-list#1152]
Nuclio and the Future of Serverless Computing<br>
http://ift.tt/2ij0lNF<br><img width="640" height="443" src="http://ift.tt/2iNZV5s" class="webfeedsFeaturedVisual wp-post-image wp-stateless-item" style="display: block; margin: auto; margin-bottom: 20px;" data-image-size="large" data-stateless-media-bucket="cdn.thenewstack.io" data-stateless-media-name="media/2017/10/60a587a1-nuclio.jpg"><a href="http://iguaz.io/blog/" class="clearfix infoBlock" target="_blank"></a>
<div class="infoBlockImageBlock"><img class="infoBlockImage" src="http://ift.tt/2mwE4wT" style="border-radius: 50%;"></div>
<div class="infoBlockTextBlock">
<div class="infoBlockTitle">Yaron Haviv</div>
<div class="infoBlockText">As the CTO and founder of Iguazio, Yaron is a serial entrepreneur who has deep technical experience in the fields of big data, cloud, storage and networking. Prior to Iguazio, Haviv was the Vice President of Datacenter Solutions at Mellanox, where he led technology innovation, software development and solution integrations. Haviv was the key driver of open source initiatives and new solutions with leading database and storage vendors, enterprise organizations, cloud and Web 2.0 customers. Before Mellanox, Haviv was the CTO and Vice President of R&amp;D at Voltaire, a high performance, computing, IO and networking company. Haviv often speaks at big data and cloud technology events. He tweets as @yaronhaviv.</div>
</div>
<p>Serverless platforms allow developers to focus on building and running auto-scaling applications without worrying about managing servers, as server provisioning and maintenance are taken care of behind the scenes. Industry demand for instant results has therefore made <a href="http://ift.tt/2mctD3P" class="local-link">serverless platforms</a> the new buzz.</p>
<p>However, serverless platforms have challenges which limit their usability and applicability:</p>
<ul>
<li>Slow performance and lack of concurrency (single threaded)</li>
<li>Lock-in to platform specific event and data sources</li>
<li>Complexity of application state maintenance, code dependencies and service dependencies</li>
<li>They are impossible to develop, debug, test and deploy in a hybrid or multicloud environment</li>
</ul><p>Latency of <a href="http://ift.tt/2wMfGvu" class="ext-link" rel="external nofollow">tens or hundreds of milliseconds</a> is the norm today when developing serverless functions in the cloud, and we&rsquo;re lucky if we manage to run more than a few thousands of events/sec without taking on a second mortgage. This limits the usage to non-performance-sensitive front-end apps or glue logic. Serverless can address many more workloads if it were faster or more efficient.</p>
<blockquote>
<p>Many serverless platforms &ldquo;added more features, but fundamental architectural decisions cannot be patched over time.&rdquo;</p>
</blockquote>
<p>We need to feed functions from messaging or streaming sources as we break past the comfort zone of simple web apps. We must also store state in cloud-specific databases/storage and use cloud-specific logging and monitoring tools. These integrations are not trivial and means that our function code is now tied to a specific cloud platform. What if we wanted to swap vendors? Or use a multicloud deployment? Or debug some code on our laptop? Or regression test on our mini cluster? Just forget about it.</p>
<h2>Yeah, But What about These Numerous Open Source Serverless/FaaS Projects&hellip;?</h2>
<p>Many open source serverless, or functions-as-a-service (FaaS), projects started as a way to rapidly develop front-end apps or glue logic. They weren&rsquo;t too focused on performance, support for a variety of event sources or better integration with state, data, logging, or versioning to make them applicable for more demanding applications or deliver an end-to-end service experience.</p>
<p>They&rsquo;ve now added more features, but fundamental architectural decisions cannot be patched over time. Some popular FaaS solutions address the challenge of triggering any containerized apps but work like CGI (every event launches the application and all its data connections from scratch), not like something that can work on apps with a high ingestion rate.</p>
<h2>So What Can We Do?</h2>
<ol>
<li>Build serverless architecture with a focus on parallelism and CPU/Mem/IO from the ground up.</li>
<li>Define a common event format across multiple event types, frameworks and cloud providers.</li>
<li>Decouple event/data sources from the function code.</li>
<li>Provide stronger data security and access control.</li>
<li>Increase the emphasis on serverless debugging, logging and instrumentation tools/features.</li>
<li>Address continuous development and operations across hybrid and multi-cloud environments. Make functions portable and versioned along with their event feed and data dependencies.</li>
</ol><p>The above criteria guided my company&nbsp;<a href="http://ift.tt/2fyoUc0" class="ext-link" rel="external nofollow">iguazio</a> when we built <a href="http://ift.tt/2vYo5gh" class="ext-link" rel="external nofollow">nuclio</a>&nbsp;&mdash; a new advanced and high-performance open source serverless framework. We know it won&rsquo;t be complete without a broader eco-system that takes on these challenges. We are actively working with the <a href="http://bit.ly/2quBaL4" class="ext-link" rel="external nofollow">Cloud Native Computing Foundation</a> to come up with industry-wide solutions and have participated in authoring this <a href="http://ift.tt/2zjSyd4" class="ext-link" rel="external nofollow">serverless paper</a> to foster collaboration.</p>
<p>Here is how <a href="http://ift.tt/2vYo5gh" class="ext-link" rel="external nofollow">nuclio</a> tackles those challenges:</p>
<h2>Serverless Performance</h2>
<p>Application performance is killed by stuff like lack of parallelism, blocked access to resources (network, io, etc.) and excessive memory allocation/deallocation/encoding.</p>
<p>Serverless functions suffer the most <a href="http://ift.tt/2iOsbVI" class="ext-link" rel="external nofollow">since they are single threaded</a>. Access to resources is blocked and, in many cases, connections restart per invocation. Serverless is also bloated with data copies, message serialization/deserialization, and context switches. No wonder performance is so slow! This impacts the type of use-cases we can serve as well as increases application costs (due to the need for more CPU and Memory resources).</p>
<p><a href="http://ift.tt/2vYo5gh" class="ext-link" rel="external nofollow">Nuclio</a> introduces the notion of lightweight worker context, as multiple function tasks run in parallel on the same resources (using Go routines). Data buffers are reused eliminating allocation/deallocation and garbage cleaning. Access to external resources/data is always non-blocking. We eliminated the need for (JSON) serialization and data copies. The communication between nuclio wrapper processes (in Go) and language-specific runtimes is non-blocking and uses shared memory.</p>
<p>We made sure incoming data or streams are partitioned and dynamically balanced across multiple function instances to gain linear scalability in cases of large-scale message handling (with nuclio&rsquo;s dealer service which can sustain failures while supporting different delivery modes/guarantees, such as &lsquo;at-least-once,&rsquo; &lsquo;at-most-once,&rsquo; etc.). The results are that nuclio can process 400,000 events per second using a single process. The bottom line is that nuclio is 100 times faster than mainstream serverless solutions.</p>
<h2>Functions Triggered by Any Type of Source</h2>
<p>When we work with <a href="http://ift.tt/2mTIdP8" class="local-link">AWS Lambda</a>, currently the most popular serverless service, we are limited to a set of platform event sources (such as HTTP gateway, Kinesis and SQS). Each event source generates a different event structure even when it carries the same payload (<a href="http://ift.tt/2iP0R9U" class="ext-link" rel="external nofollow">forcing hacks like this</a>). This means our function code is locked-in to a specific event implementation and provider.</p>
<p>Nuclio introduces the notion of a common event structure, separating protocols which are specific to the event header from content in the body, and supporting various serialization options (not forcing expensive JSON deserialization). With nuclio, the same function code can be triggered by HTTP, <a href="http://ift.tt/1nhMZhV" class="ext-link" rel="external nofollow">Kinesis</a>,&nbsp;<a href="http://ift.tt/18pqP5X" class="ext-link" rel="external nofollow">Kafka</a>,&nbsp;<a href="http://ift.tt/1hBqVI8" class="ext-link" rel="external nofollow">RabbitMQ</a>,&nbsp;<a href="http://mqtt.org/" class="ext-link" rel="external nofollow">MQTT</a>,&nbsp;<a href="https://nats.io/" class="ext-link" rel="external nofollow">NATS</a>,&nbsp;<a href="http://ift.tt/2gsAx0P" class="ext-link" rel="external nofollow">iguazio&rsquo;s V3IO</a>, file content, or an event emulator.</p>
<p>We can now develop our functions with an emulator, test with HTTP, use a stream for production and even change the source if you move between cloud providers or beta/production setups.</p>
<p>Nuclio&rsquo;s common event source approach was proposed as a <a href="http://ift.tt/2zjSz0C" class="ext-link" rel="external nofollow">CNCF specification</a> to address the problem of serverless lock-in and complexity. Hopefully, multiple platforms will adopt this concept.</p>
<h2>Accelerating, Securing and Abstracting Function Access to Various Data Sources</h2>
<p>Functions are stateless. They access external data services frequently to retrieve state or content, store results and pass messages. Today, specific data access implementation and credentials are usually either hard coded into the function, or at best, they use some environment variables. This means that we cannot build a function that will be portable and it is very hard to use the same function with different data types or repositories during development, testing and production.</p>
<p>Nuclio provides pluggable data bindings to functions. Developers work against data binding and operators can decide at deployment time which data systems and datasets are used. This concept, originally pioneered by Azure, results in:</p>
<ul>
<li>Simplicity: coding against simple APIs without taking care of SDKs and connection details.</li>
<li>Security: credentials are not part of the function code/variables.</li>
<li>Portability: move functions around and work with different data sources.</li>
<li>Reusability: the same function can be used in multiple use-cases by changing the bound datasets.</li>
<li>Performance: data and connections are cached across invocations and work with non-blocking IO and zero-copy access.</li>
</ul><h2>Simplifying Function Debugging, Logging and Instrumentation</h2>
<p>Debugging serverless can be frustrating, as it has no tracing or breakpoints. How can we automate regression testing, especially when we&rsquo;re tied to external event and data sources?</p>
<p>Logging and prints all go to unstructured and resource-intensive logs with no way to control the detail level vs overhead tradeoffs. Automation requires complex log parsing.</p>
<p>There is still a lot to do in that space and it&rsquo;s a challenge that needs to be addressed through collaboration and standardization between serverless platform providers and software providers of IDEs, testing, instrumentation tools and the like. nuclio takes a few important steps in the right direction:</p>
<ul>
<li>Built-in non-intrusive structured and unstructured logging to a range of output options (screen, HTTP, file, log streams) with dynamically controlled verbosity level. This allows us to place many log/debug points in our code and change verbosity levels on the fly or per invocation to diagnose bugs and failures without changing the code. This structure option makes it ideal for test automation and analysis.</li>
<li>Nuclio functions import the nuclio-SDK, making our local IDE work as if it was a local program, with auto-completions and breakpoints. The SDK provides a way to run our function locally, print logs to the screen/file, feed it from emulated events and emulate function input/outputs with pluggable data bindings.</li>
<li>For regression testing, we can create an event feed emulation (from files/http/streams/..) with an emulated input data set and compare the logs and data output with expected results, all without changing code.</li>
<li>Additional debugging and instrumentation features are in our roadmap, including custom statistics and <a href="http://ift.tt/2h1MIoK" class="local-link">OpenTracing</a> integration.</li>
</ul><h2>Delivering Hybrid and Multicloud CI/CD Pipelines and Distributed Deployment Models</h2>
<p>For some, doing all the work on a single cloud is a viable option, but for most others, that is expensive and limiting. A typical case involves developers working on laptops, regression testing on dedicated clusters that get bombarded with events 24&times;7 (not ideal for the serverless per invocation cost model), beta and production deployments. In the Internet of Things or edge scenarios, these same functions may also need to deploy in multiple edge locations. This means we need a versioned and distributed development pipeline, with an option for rolling or canary upgrades (deploy an 80/20 version mix).</p>
<p>Nuclio features mentioned above&nbsp;&mdash; such as the ability to run anywhere, the abstraction of events, logs, and data and the logging capabilities &mdash; are important features which enable deployment in multiple locations. In addition, nuclio provides a federated and versioned deployment approach. Compiled functions generate an artifact (binary or container image) that can be stored locally or on a shared image repo, function can have multiple versions each with unique metadata, events and data. With nuclio, we can choose to build/test on one cluster and deploy the same artifact on multiple clusters without rebuilding our code (deploy from image).</p>
<h2>Summary</h2>
<p>Serverless is a young and fast-growing space. It is a very promising way to accelerate businesses and it still has lots of room for innovation. I believe this can only be achieved through better collaboration and by the various vendors and members in the community sharing between themselves. We hope others will join the CNCF&rsquo;s development and standardization efforts in the space.</p>
<p>If you want to play with a stand-alone nuclio version, it&rsquo;s as simple as copy/pasting this line into your command line interface:</p>
<p>docker run -p 8070:8070 -v /var/run/docker.sock:/var/run/docker.sock nuclio/playground\</p>
<p>Open a browser at HTTP://&lt;machine-ip&gt;:8070. This will bring up the nuclio playground. Select an example function, edit, deploy and invoke. While you wait, please give it a star. Visit our <a href="http://ift.tt/2vYo5gh" class="ext-link" rel="external nofollow">GitHub page</a>&nbsp;for the full version and more details.</p>
<p class="attribution">The <a href="http://bit.ly/2quBaL4" class="ext-link" rel="external nofollow">Cloud Native Computing Foundation</a>&nbsp;is a sponsor of The New Stack.</p>
<p>The post <a rel="nofollow" href="http://ift.tt/2ij0lNF">Nuclio and the Future of Serverless Computing</a> appeared first on <a rel="nofollow" href="https://thenewstack.io">The New Stack</a>.</p>
<br><br>
via Blog &ndash; The New Stack https://thenewstack.io<br>
October 31, 2017 at 06:00AM
